### PR TITLE
RLLC v1

### PR DESCRIPTION
**Pull request by svwingerden**
This PR implements a version of weight-restricted LLC estimation. Key functionalities:
1. Fixed `sample` to work when only parts of a model have `requires_grad = True` (easiest method of RLLC estimation)
2. Added in code that takes in a dict of `{'model_param_name_string': tensor_shaped_like_named_model_param_weights}`, and allows for LLC estimates using this boolean mask. This allows freezing specific weights (rather than entire layers, as in 1.).

Notebooks are not yet implemented, as this is primarily an internal research tool rather than a finished product. See `tests/optim/rllc_test.py` for some examples.

### CODE REVIEW

File: src/devinterp/slt/sampler.py [hunk 1]
```diff
@@ -87,6 +102,9 @@ def sample_single_chain(
         loss = criterion(y_preds, ys)
 
         loss.backward()
+        # if optimize_over_per_model_param: # not sure this is needed tbh
+        #     for param_name, optimize_over in optimize_over_per_model_param.items():
+        #         getattr(model, param_name).grad = getattr(model, param_name).grad*optimize_over
```
Comment (georgeyw) [inline src/devinterp/slt/sampler.py:107]: if this isn't needed, delete comment?

File: src/devinterp/optim/sgld.py [hunk 2]
```diff
@@ -71,6 +71,7 @@ def __init__(
         temperature: Union[Callable, float] = 1.0,
         bounding_box_size=None,
         save_noise=False,
+        optimize_over=None,
```
Comment (georgeyw) [inline src/devinterp/optim/sgld.py:74]: i think this parameter is no longer used and instead this is in `params` which is updated by the sampler? if so we can delete it (and from the default params below)

File: src/devinterp/optim/sgld.py [hunk 3]
```diff
@@ -89,6 +90,7 @@ def __init__(
             localization=localization,
             temperature=temperature,
             bounding_box_size=bounding_box_size,
+            optimize_over=optimize_over,
```
Comment (georgeyw) [inline src/devinterp/optim/sgld.py:93]: see above comment

File: tests/optim/rllc_test.py
Comment (georgeyw) [inline tests/optim/rllc_test.py:1]: i'm confused about these tests. it reads to me like we're taking different models and either applying LLC or weight RLLC to them and estimating LLCs, then checking that the estimate is different. but wouldn't this happen anyways if the models were different, even if RLLC wasn't working?

Comment (georgeyw) [reply tests/optim/rllc_test.py:1]: oh i see, sorry i misread i think. would it be too much trouble to add another test that does full LLC and wRLLC on the same model and checks that the values are different?
